compiler/rustc_const_eval/src/const_eval/panic.rs

```diff
@@ -0,0 +1,262 @@
+use rustc_hir::def_id::DefId;
+use rustc_middle::ty::{self, layout::LayoutOf, subst::Subst};
+use rustc_span::sym;
+use std::cell::Cell;
+use std::fmt::{self, Debug, Formatter};
+
+use crate::interpret::{FnVal, InterpCx, InterpErrorInfo, InterpResult, OpTy};
+
+use super::CompileTimeInterpreter;
+
+struct Arg<'mir, 'tcx, 'err> {
+    cx: &'err InterpCx<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>>,
+    arg: OpTy<'tcx>,
+    fmt_trait: DefId,
+    err: &'err Cell<Option<InterpErrorInfo<'tcx>>>,
+}
+
+impl Debug for Arg<'_, '_, '_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self.cx.fmt_arg(self.arg, self.fmt_trait, f) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                self.err.set(Some(e));
+                Err(fmt::Error)
+            }
+        }
+    }
+}
+
+impl<'mir, 'tcx> InterpCx<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>> {
+    fn fmt_arg(
+        &self,
+        arg: OpTy<'tcx>,
+        fmt_trait: DefId,
+        f: &mut Formatter<'_>,
+    ) -> InterpResult<'tcx> {
+        let fmt_trait_sym = self.tcx.item_name(fmt_trait);
+        let fmt_trait_name = fmt_trait_sym.as_str();
+
+        macro_rules! dispatch_fmt {
+            ($e: expr, $($t: ident)|*) => {
+                let _ = match &*fmt_trait_name {
+                    $(stringify!($t) => fmt::$t::fmt($e, f),)*
+                    _ => Debug::fmt($e, f),
```

Comment:
Could you add a comment explaining what this macros does? I cannot make sense of it.

Also please not explicitly that we are dispatching to the standard library formatting machinery here, so we are guaranteed that this stays in sync with the logic in the standard library -- is that correct?

Reply:
This just generates match arms like `"Display" => fmt::Display::fmt(e, f),`.

Yes, it's correct that this would mean the formatting result to be in sync with std. Though I don't think we want to guarantee  that to the user.

Reply:
It seems unfortunate to do string comparisons on the trait name. I wonder if it would be better to make Display etc. lang items instead?

Reply:
They certainly shouldn't be lang items but could be made into diagnostic items. With that said, `format_args!` expansion in `rustc_builtin_macros` is using strings to name these as well, so I don't think it's a big deal.

Reply:
> Yes, it's correct that this would mean the formatting result to be in sync with std. Though I don't think we want to guarantee that to the user.

I think we have to make this guarantee. All this machinery should be entirely invisible, it just works around limitations in the main CTFE machinery.

Reply:
Also, instead of having a `Debug` fallback case, shouldn't this `bug!` in case of an unknown trait name, to ensure we do dispatch to the right code?